packages/compartment-mapper/src/compartment-map.js

```diff
@@ -0,0 +1,418 @@
+    `${path} must not have extra properties, got ${q(
+      Object.keys(extra),
+    )} in ${q(url)}`,
+  );
+  assert.typeof(
+    exit,
+    'string',
+    `${path}.exit must be a string, got ${q(exit)} in ${q(url)}`,
+  );
+};
+
+/**
+ * @param {unknown} allegedModule
+ * @param {string} path
+ * @param {string} url
+ */
+const assertModule = (allegedModule, path, url) => {
+  const moduleDescriptor = Object(allegedModule);
+  assert(
+    allegedModule === moduleDescriptor && !Array.isArray(moduleDescriptor),
+    `${path} must be an object, got ${allegedModule} in ${q(url)}`,
+  );
+
+  const { compartment, module, location, parser, exit } = moduleDescriptor;
+  if (compartment !== undefined || module !== undefined) {
+    assertCompartmentModule(moduleDescriptor, path, url);
+  } else if (location !== undefined || parser !== undefined) {
+    assertFileModule(moduleDescriptor, path, url);
+  } else if (exit !== undefined) {
+    assertExitModule(moduleDescriptor, path, url);
+  } else {
+    assert.fail(
+      `${path} is not a valid module descriptor, got ${q(allegedModule)} in ${q(
+        url,
+      )}`,
+    );
+  }
+};
+
+/**
+ * @param {unknown} allegedModules
+ * @param {string} path
+ * @param {string} url
+ */
+const assertModules = (allegedModules, path, url) => {
+  const modules = Object(allegedModules);
+  assert(
+    allegedModules === modules || !Array.isArray(modules),
+    `modules must be an object, got ${allegedModules} in ${q(url)}`,
+  );
+  for (const [key, value] of Object.entries(modules)) {
+    assertModule(value, `${path}.modules[${q(key)}]`, url);
+  }
+};
+
+/**
+ * @param {unknown} allegedParsers
+ * @param {string} path
+ * @param {string} url
+ */
+const assertParsers = (allegedParsers, path, url) => {
+  if (allegedParsers === undefined) {
+    return;
+  }
+  const parsers = Object(allegedParsers);
+  assert(
+    allegedParsers === parsers && !Array.isArray(parsers),
+    `${path}.parsers must be an object, got ${allegedParsers} in ${q(url)}`,
+  );
+
+  for (const [key, value] of Object.entries(parsers)) {
+    assert.typeof(
+      key,
+      'string',
+      `all keys of ${path}.parsers must be strings, got ${key} in ${q(url)}`,
+    );
+    assert.typeof(
+      value,
+      'string',
+      `${path}.parsers[${q(key)}] must be a string, got ${value} in ${q(url)}`,
+    );
+    assert(
+      moduleLanguages.includes(value),
+      `${path}.parsers[${q(key)}] must be one of ${q(
+        moduleLanguages,
+      )}, got ${value} in ${q(url)}`,
+    );
+  }
+};
+
+/**
+ * @param {unknown} allegedScope
+ * @param {string} path
+ * @param {string} url
+ */
+const assertScope = (allegedScope, path, url) => {
+  const scope = Object(allegedScope);
+  assert(
+    allegedScope === scope && !Array.isArray(scope),
+    `${path} must be an object, got ${allegedScope} in ${q(url)}`,
+  );
+
+  const { compartment, ...extra } = scope;
+  assert(
+    Object.keys(extra).length === 0,
```

Comment:
I'm tempted to suggest this `assertNoExtra(extra, explanation)` pattern be factored out so the calls fit on a single line, but I think I'm just feeling grumpy towards `prettier`'s verbosity today. The work this file is doing is completely straightforward.